views.go

```diff
@@ -391,6 +398,87 @@ func (lk *logKeeper) viewAllLogs(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (lk *logKeeper) viewTestInDatabase(r *http.Request, buildID string, testID string) (*logFetchResponse, *apiError) {
+	build, err := model.FindBuildById(buildID)
+	if err != nil || build == nil {
+		return nil, &apiError{Err: "view test by id: build not found", code: http.StatusNotFound}
+	}
+
+	test, err := model.FindTestByID(testID)
+	if err != nil || test == nil {
+		return nil, &apiError{Err: "test not found"}
+	}
+
+	logsChan, err := model.MergedTestLogs(test)
+	if err != nil {
+		lk.logErrorf(r, "Error finding global logs during test: %v", err)
+		return nil, &apiError{Err: err.Error(), code: http.StatusInternalServerError}
+	}
+
+	result := logFetchResponse{
+		logLines: logsChan,
+		build:    build,
+		test:     test,
+	}
+	return &result, nil
+}
+
+func (lk *logKeeper) viewTestInS3(r *http.Request, buildID string, testID string) (*logFetchResponse, *apiError) {
+	var build *model.Build
+	var buildErr error
+	wg := sync.WaitGroup{}
+
+	wg.Add(3)
+	go func() {
+		defer wg.Done()
+		build, buildErr = lk.opts.Bucket.FindBuildByID(r.Context(), buildID)
+	}()
+
+	var test *model.Test
+	var testErr error
+	go func() {
+		defer wg.Done()
+		test, testErr = lk.opts.Bucket.FindTestByID(r.Context(), buildID, testID)
+	}()
+
+	var logsChan chan *model.LogLineItem
+	var logsChanErr error
+	go func() {
+		defer wg.Done()
+		logsChan, logsChanErr = lk.opts.Bucket.GetTestLogLines(r.Context(), buildID, testID)
+	}()
+
+	wg.Wait()
+
+	if buildErr != nil {
+		return nil, &apiError{Err: "view test in S3 by id: error fetching build", code: http.StatusInternalServerError}
```

Comment:
In the `StatusInternalServerError` cases I see why we don't want to tell the client what the error was, but does it make sense to log it? We could pass it to `lk.logErrorf`.

Reply:
So you're saying we think we shouldn't return StatusInternalServerError?

Reply:
I'm with you on returning `StatusInternalServerError` to the client. I was just suggesting that we log the actual error in `buildErr` to Splunk or something. This way if S3 starts returning a particular error, we'll be able to see it in Splunk.

Reply:
Updated these to log the error, but not return that result to the user.

Reply:
Happy to change this again if I interpreted this wrong though.

Reply:
I'd been thinking more along the lines of 
```go
if buildErr != nil {
	lk.logErrorf(r, "error fetching build: %v", buildErr)
	return nil, &apiError{Err: "error fetching build", code: http.StatusInternalServerError}
if build == nil {
	return nil, &apiError{Err: fmt.Sprintf("no matching build for id '%s'", buildID), code: http.StatusNotFound}
}
```
This way let's say AWS returns an error we'll be logging the actual error.